buildstock_query/aggregate_query.py

```diff
@@ -305,13 +435,230 @@ def get_lower_timestamps(day, hour):
 
         batch_id = self._bsq.submit_batch_query(query_strs)
         if exact_times:
-            vals, = self._bsq.get_batch_query_result(batch_id, combine=False)
+            (vals,) = self._bsq.get_batch_query_result(batch_id, combine=False)
             return vals
         else:
             lower_vals, upper_vals = self._bsq.get_batch_query_result(batch_id, combine=False)
-            avg_upper_weight = np.mean([min_of_hour / sim_interval_seconds for hour in at_hour if
-                                        (min_of_hour := hour * 3600 % sim_interval_seconds)])
+            avg_upper_weight = np.mean(
+                [
+                    min_of_hour / sim_interval_seconds
+                    for hour in at_hour
+                    if (min_of_hour := hour * 3600 % sim_interval_seconds)
+                ]
+            )
             avg_lower_weight = 1 - avg_upper_weight
             # modify the lower vals to make it weighted average of upper and lower vals
             lower_vals[enduses] = lower_vals[enduses] * avg_lower_weight + upper_vals[enduses] * avg_upper_weight
             return lower_vals
+
+    def validate_partition_by(self, partition_by: Sequence[str]):
+        if not partition_by:
+            return []
+        [self._bsq._get_gcol(col) for col in partition_by]  # making sure all entries are valid
+        return partition_by
+
+    @gather_params(Query)
+    def _query(
```

Comment:
This is the core of the new implementation, right? There's a lot of logic here. Suggest breaking it up into bite-size chunks for readability, maybe when savings_shape() get removed.

e.g., one function just to get the upgrade_col seems warranted, as well as one function to do the final execution when all the query pieces are generated.